tests/unit/recommenders/evaluation/test_python_evaluation.py

```diff
@@ -75,6 +76,29 @@ def rating_nohit():
             DEFAULT_PREDICTION_COL: [12, 14, 13, 12, 11, 10, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 14, 13],
         }
     )
+
+
+@pytest.fixture
+def top_3_items_true():
+    # Both are correct top 3 items of rating_true
+    return [
+        pd.DataFrame(
+            {
+                DEFAULT_USER_COL: [1, 1, 1, 2, 2, 2, 3, 3, 3],
+                DEFAULT_ITEM_COL: [1, 2, 3, 1, 4, 5, 2, 5, 6],
+                DEFAULT_RATING_COL: [5, 4, 3, 5, 5, 3, 5, 5, 5],
+                "rank": [1, 2, 3, 1, 2, 3, 1, 2, 3],
+            }
+        ),
+        pd.DataFrame(
+            {
+                DEFAULT_USER_COL: [1, 1, 1, 2, 2, 2, 3, 3, 3],
+                DEFAULT_ITEM_COL: [1, 2, 3, 1, 4, 6, 2, 5, 6],
+                DEFAULT_RATING_COL: [5, 4, 3, 5, 5, 3, 5, 5, 5],
+                "rank": [1, 2, 3, 1, 2, 3, 1, 2, 3],
+            }
+        ),
+    ]
```

Comment:
This is good @ChuyangKe, some questions. 

The rank list is always [1,2,3, 1,2,3...] in that case, why do we need it?

In the assert you have an or clause because user 2 can get items 1,4,5 or 1,4,6, are you sure that you can get 1,4,6? can you explain how the ordering is done? and also, in this context, how the ordering is done with the rating is the same, for example, user 3 has items 2,5,6 with 5 stars. In which case the rank swaps, for example, getting items 6,5,2?

Reply:
This rank list is the expected output of get_top_k_items() 

> "Returns: as.DataFrame: DataFrame of top k items for each user, sorted by `col_user` and `rank`

For user 2, both 1,4,5 and 1,4,6 are correct. This is because item 1,4 have the highest score of 5, and after that item 5 and 6 have the second highest score of 3, so either one should be a valid choice.

What is missing, is the permutation of 2,5,6 for user 3. All cases (2,5,6; 2,6,5; 5,2,6; 5,6,2; 6,2,5; 6,5,2) should be accepted following the same rationale. I will submit a change for this.

```
def rating_true():
    return pd.DataFrame(
        {
            DEFAULT_USER_COL: [1, 2, 2, 2, 2, 2, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 1, 1],
            DEFAULT_ITEM_COL: [3, 1, 4, 5, 6, 7, 2, 5, 6, 8, 9, 10, 11, 12, 13, 14, 1, 2],
            DEFAULT_RATING_COL: [3, 5, 5, 3, 3, 1, 5, 5, 5, 4, 4, 3, 3, 3, 2, 1, 5, 4],
        }
    )
```